File: dotnet/ThirtyFour.UI.Tests/Window/DefaultMatcherTests.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using ThirtyFour.UI.Window;

namespace ThirtyFour.UI.Tests.Window
{
    [TestClass]
    public class DefaultMatcherTests : BaseTestSuite
    {

        [TestMethod]
        public void ExceptionThrownWhenNoPopupWindowsExist()
        {
            try
            {
                new DefaultMatcher().MatchWindow(driver, 4);
                Assert.Fail("An exception should have been thrown");
            }
            catch (Exception e)
            {
                Assert.AreEqual("Unable to find a new Window", e.Message);
            }
        }

        [TestMethod]
        public void WindowSwitchedToCorrectWindow()
        {

            driver.FindElement(By.LinkText("Popup")).Click();
            new DefaultMatcher().MatchWindow(driver, 4);
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(SeleniumExtras.WaitHelpers.ExpectedConditions.UrlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));
        }

        [TestMethod]
        public void WindowSwitchedWhenMatchingDelayedWindow()
        {
            driver.FindElement(By.LinkText("Delayed Popup")).Click();
            new DefaultMatcher().MatchWindow(driver, 5);
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(SeleniumExtras.WaitHelpers.ExpectedConditions.UrlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));
          
        }
    }
}

File: java/src/test/java/com/github/danhumphrey/thirtyfour/ui/test/window/DefaultMatcherTests.java
package com.github.danhumphrey.thirtyfour.ui.test.window;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import com.github.danhumphrey.thirtyfour.ui.test.BaseTestSuite;
import com.github.danhumphrey.thirtyfour.ui.window.DefaultMatcher;

public class DefaultMatcherTests extends BaseTestSuite {
	
	@Test
	public void testExceptionThrownWhenNoPopupExists() throws Exception {
		
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		try {
			new DefaultMatcher().matchWindow(driver, 2000);
			Assertions.fail("Expected exception not thrown");
		} catch(Exception ex) {
		} finally {
			quitExtraDrivers(currentHandle);
		}
	}
	
	
	@Test
	public void testWindowSwitchesToCorrectWindow() throws Exception {
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		driver.findElement(By.linkText("Popup")).click();
		
		try {
			new DefaultMatcher().matchWindow(driver, 2000);
	        new WebDriverWait(driver, 2).until(ExpectedConditions.urlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));
			
		} finally {
			quitExtraDrivers(currentHandle);
		}
    }
	
	@Test
	public void testWindowSwitchesToCorrectWindowWhenDelayed() throws Exception {
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		driver.findElement(By.linkText("Delayed Popup")).click();
		
		try {
			new DefaultMatcher().matchWindow(driver, 5000);
	        new WebDriverWait(driver, 2).until(ExpectedConditions.urlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));
			
		} finally {
			quitExtraDrivers(currentHandle);
		}
    }
	
}
